config/script.go

```diff
@@ -99,102 +146,140 @@ func evalScriptDesc(evalctx *eval.Context, expr hhcl.Expression, name string) (s
+	it := cmdList.ElementIterator()
 	for it.Next() {
 		index++
 		_, elem := it.Element()
-		if elem.Type() != cty.String {
-			errs.Append(errors.E(ErrScriptInvalidTypeCommand, expr.Range(),
-				"field %s must be a list(string) but element %d has type %q",
-				name, index, elem.Type().FriendlyName()))
+		if !elem.Type().IsTupleType() {
+			errs.Append(errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
+				"commands must be a list of list, but element %d has type %q",
+				index, elem.Type().FriendlyName()))
 			continue
 		}
-		evaluatedCommand = append(evaluatedCommand, elem.AsString())
+
+		if elem.LengthInt() == 0 {
+			errs.Append(errors.E(ErrScriptEmptyCmds, expr.Range(), "commands item %d is empty", index))
+			continue
+		}
+
+		evaluatedCommand, err := unmarshalScriptJobCommand(elem, expr)
+		if err != nil {
+			errs.Append(err)
+			continue
+		}
+
+		evaluatedCommands = append(evaluatedCommands, evaluatedCommand)
 	}
 
 	if err := errs.AsError(); err != nil {
 		return nil, err
 	}
 
-	return evaluatedCommand, nil
+	return evaluatedCommands, nil
 }
 
-func evalScriptJobCommands(evalctx *eval.Context, expr hhcl.Expression, name string) ([][]string, error) {
-	val, err := evalctx.Eval(expr)
-	if err != nil {
-		return nil, errors.E(ErrScriptSchema, expr.Range(),
-			err, "evaluating %s", name)
-	}
-
-	if !val.Type().IsTupleType() {
-		return nil, errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
-			"%s should be a list(string) type", name)
+func unmarshalScriptJobCommand(cmdValues cty.Value, expr hhcl.Expression) (*ScriptCmd, error) {
+	if !cmdValues.Type().IsTupleType() {
+		return nil, errors.E(ErrScriptInvalidTypeCommand, expr.Range(), "command must be a list")
 	}
 
-	if val.LengthInt() == 0 {
+	if cmdValues.LengthInt() == 0 {
 		return nil, errors.E(ErrScriptEmptyCmds, expr.Range())
 	}
 
 	errs := errors.L()
-	evaluatedCommands := [][]string{}
+	r := &ScriptCmd{}
 
-	index := -1
-	it := val.ElementIterator()
+	index := 0
+	lastIndex := cmdValues.LengthInt() - 1
+
+	it := cmdValues.ElementIterator()
 	for it.Next() {
-		index++
 		_, elem := it.Element()
-		if !elem.Type().IsTupleType() {
-			errs.Append(errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
-				"field %s must be a list of list(string) but element %d has type %q",
-				name, index, elem.Type().FriendlyName()))
-			continue
+		if elem.Type() == cty.String {
+			r.Args = append(r.Args, elem.AsString())
+		} else if index == lastIndex {
+			if elem.Type().IsObjectType() {
+				var err error
+				r.Options, err = unmarshalScriptCommandOptions(elem, expr)
+				errs.Append(err)
+			} else {
+				errs.Append(errors.E(ErrScriptInvalidTypeCommand, expr.Range(),
+					"command options must be an object, but last element has type %s",
+					elem.Type().FriendlyName()))
+			}
+		} else {
+			errs.Append(errors.E(ErrScriptInvalidTypeCommand, expr.Range(),
+				"command must be a list(string), but element %d has type %s",
+				index, elem.Type().FriendlyName()))
 		}
```

Comment:
This if-else chain is a bit hard to read because of the nested clauses. We could use a switch statement here to get rid of the nested clauses? (and also make it easier to add more cases in future) e.g.:

```golang
	for it.Next() {
		_, elem := it.Element()
		switch {
		case elem.Type() == cty.String:
			r.Args = append(r.Args, elem.AsString())
		case (index == lastIndex) && elem.Type().IsObjectType():
			var err error
			r.Options, err = unmarshalScriptCommandOptions(elem, expr)
			errs.Append(err)
		case (index == lastIndex) && !elem.Type().IsObjectType():
			errs.Append(errors.E(ErrScriptInvalidTypeCommand, expr.Range(),
				"command options must be an object, but last element has type %s",
				elem.Type().FriendlyName()))
		default:
			errs.Append(errors.E(ErrScriptInvalidTypeCommand, expr.Range(),
				"command must be a list(string), but element %d has type %s",
				index, elem.Type().FriendlyName()))
		}
		index++
	}
```

What do you think?

Reply:
I disagree on this one, I prefer the nested version because the decision tree is clearer like that.

For example, when I look at this, it's obvious in which branch I end up:
```
if a {
  if b {
  } else {
  }
} else {
}
```
As opposed to
```
if a && b {
} else if a && !b {
} else {
}
```
I look at this, I will wonder why the conditions are formulated like this and whether I'm missing any case.